### PR TITLE
fix(metrics): Handle properly the intervals length calculation

### PR DESCRIPTION
**Pull request by iambriccardo**
This PR aims at implementing the proper logic for avoiding the case in which a division by zero is performed. The case happens when the integer representation of two numbers is equal, thus resulting in their difference being = 0.

In order to avoid this issue entirely we require that the difference between `start` and `end` is at least equal to the `interval` which mathematically means that the resulting division will be >= 1, guaranteeing that `int()` rounding will result in two different numbers.

### CODE REVIEW

File: src/sentry/snuba/metrics/query.py [hunk 1]
```diff
@@ -257,16 +257,28 @@ def calculate_intervals_len(
             denominator = granularity
         else:
             assert start is not None and interval > 0
+
+            start_in_seconds = start.timestamp()
+            end_in_seconds = end.timestamp()
+            # This condition is required because the formatting of start and end uses the "int()" type conversion
```
Comment (andriisoldatenko) [inline src/sentry/snuba/metrics/query.py:263]:
```suggestion
            # This condition is required because the formatting of `start` and `end` uses the "int()" function to convert
```

File: src/sentry/snuba/metrics/query.py [hunk 2]
```diff
@@ -257,16 +257,28 @@ def calculate_intervals_len(
...
+            # which automatically cuts off any decimal digits resulting in certain cases in which end - start = 0. In
```
Comment (andriisoldatenko) [inline src/sentry/snuba/metrics/query.py:264]:
```suggestion
            # which automatically cuts off any decimal digits resulting in certain cases in which `end - start = 0`. In
```

File: src/sentry/snuba/metrics/query.py [hunk 3]
```diff
@@ -257,16 +257,28 @@ def calculate_intervals_len(
...
+            # order to avoid this problem entirely we must make sure that the integer value of
+            # "start / interval" and "end / interval" differ by at least 1. We can model it mathematically as: Given "x"
+            # amount of seconds as start time, "z" amount of seconds as end time and an interval in seconds "y" we
```
[the remark this answers is not in the document]
Comment (iambriccardo) [reply src/sentry/snuba/metrics/query.py:267]: Thanks!

File: tests/sentry/snuba/metrics/test_query.py [hunk 4]
```diff
@@ -617,15 +617,15 @@ def test_ensure_interval_set_to_granularity_in_performance_queries():
...
-@freeze_time("2022-11-03 10:10:00")
+@freeze_time("2022-11-03 10:30:00")
 @pytest.mark.parametrize(
     "granularity, interval, expected_granularity",
     [
         pytest.param(
             86400,
-            7200,
             3600,
-            id="day granularity with 2 hour interval",
```
Comment (ahmedetefy) [inline tests/sentry/snuba/metrics/test_query.py:628]: Why did you remove this test?

Comment (iambriccardo) [reply tests/sentry/snuba/metrics/test_query.py:628]: I did it because it is not anymore valid as the interval (7200) is bigger than the difference in start and end time. We could solve this by using different start and end or like i did testing with 3600 which is >= than end - start.

Comment (ahmedetefy) [reply tests/sentry/snuba/metrics/test_query.py:628]: I would not remove it, I would just create start and end with difference of 4 hrs, and keep the interval at 2 hr

Comment (iambriccardo) [reply tests/sentry/snuba/metrics/test_query.py:628]: Sure!

**Review by andriisoldatenko — APPROVED** [review]
